drivers/npmx/npmx_shell.c

```diff
@@ -82,28 +85,203 @@ typedef struct {
 		npmx_charger_t const *p_instance,
 		uint32_t resistance); /* Pointer to function setting NTC resistance for temperature treshold. */
 } temp_recalculate_data_t;
+/** @brief Supported shell argument types. */
+typedef enum {
+	SHELL_ARG_TYPE_INT_VALUE, /* Shell int32_t type used for values. */
+	SHELL_ARG_TYPE_UINT_VALUE, /* Shell uint32_t type used for values. */
+	SHELL_ARG_TYPE_BOOL_VALUE, /* Shell bool type used for values. */
+	SHELL_ARG_TYPE_UINT_INDEX, /* Shell uint32_t type used for peripherals index. */
+} shell_arg_type_t;
+
+/** @brief Shell argument result value. */
+typedef union {
+	int32_t ivalue; /* int32_t result. */
+	uint32_t uvalue; /* uint32_t result. */
+	bool bvalue; /* bool result. */
+} shell_arg_result_t;
+
+/** @brief Shell argument structure. */
+typedef struct {
+	const shell_arg_type_t type; /* Shell argument type. */
+	const char *name; /* Shell argument name. */
+	shell_arg_result_t result; /* Shell argument result. */
+} shell_arg_t;
+
+/** @brief Shell arguments infomation structure. */
+typedef struct {
+	const int expected_args; /* Expected arguments count. */
+	shell_arg_t arg[SHELL_ARG_MAX_COUNT]; /* Shell arguments. */
+} args_info_t;
+
+/** @brief Supported unit types. */
+typedef enum {
+	UNIT_TYPE_AMPERAS, /* Unit type mA. */
+	UNIT_TYPE_VOLTS, /* Unit type mV. */
+	UNIT_TYPE_CELSIUS, /* Unit type *C. */
+	UNIT_TYPE_OHMS, /* Unit type ohms. */
+	UNIT_TYPE_PCT, /* Unit type %. */
+	UNIT_TYPE_NONE, /* No unit type. */
+} unit_type_t;
 
 static const struct device *pmic_dev = DEVICE_DT_GET(DT_NODELABEL(npm_0));
 
-static bool check_error_code(const struct shell *shell, npmx_error_t err_code)
+static const char *name_get(shell_arg_type_t arg_type)
 {
-	switch (err_code) {
-	case NPMX_SUCCESS:
-		return true;
-	case NPMX_ERROR_INVALID_PARAM: {
-		shell_error(shell, "Error: invalid parameter for npmx function.");
-		return false;
+	switch (arg_type) {
+	case SHELL_ARG_TYPE_INT_VALUE:
+	case SHELL_ARG_TYPE_UINT_VALUE:
+	case SHELL_ARG_TYPE_BOOL_VALUE:
+		return "value";
+	case SHELL_ARG_TYPE_UINT_INDEX:
+		return "instance index";
+	default:
+		return "";
 	}
-	case NPMX_ERROR_IO: {
-		shell_error(shell, "Error: IO error.");
+}
+
+static bool arguments_check(const struct shell *shell, size_t argc, char **argv,
+			    args_info_t *args_info)
+{
+	assert(args_info);
+	assert(args_info->expected_args <= SHELL_ARG_MAX_COUNT);
+
+	/* argc and argv have also an information about command name, here it should be skipped. */
+	int command_arguments_count = argc - 1;
+	/* Only arguments passed to commands are used. Command name is skipped. */
+	argv++;
+
+	/* Check if all argument are passed. */
+	if (command_arguments_count < args_info->expected_args) {
+		shell_arg_t *missing_arg = &args_info->arg[command_arguments_count];
+		shell_error(shell, "Error: missing %s %s.", missing_arg->name,
+			    name_get(missing_arg->type));
```

Comment:
I have one problem with this. What if more that one argument is missing? For example, buck instance number and new buck state. Right now, it is printed, that instance AND buck state are missing. Here, it will print, that only buck instance is missing. Any idea, how to print all missing values and not just the first one?

Reply:
Bump. Maybe it should be changed to for loop iterating over all missing arguments? Or a separate handling/`shell_error` format for different `args_info->expected_args -  argc` values? Since `SHELL_ARG_MAX_COUNT ` is equal to 3, there would only be three cases